### ISSUE TITLE
Testing

### ISSUE THREAD

**Issue by Michael-Li-Bryant** [kilianzepf/conditioned_uncertain_segmentation#1]
Hi Kilian, thanks for your great job. I'm quite interested in the label style, too. I'm wondering how to test the trained model. As I'm a newer in the field, could you please share your testing code for SSN. That'll help me a lot. Thank you.@kilianzepf

**Comment by kilianzepf** [comment]
Hi @Michael-Li-Bryant, thanks for having a look at the repo! You can evaluate the conditioned SSN with the following functions: 

`def evaluate(
    model, test_loader, treshold_for_testing, metadata, forward_passes, W=True
):
    if os.path.isfile(f"saved_models/{meta.directory_name}/{model.name}.pt"):
        transfer_dict = torch.load(
            f"saved_models/{meta.directory_name}/{model.name}.pt"
        )
        model = transfer_dict["model"]
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

        model.eval()
        sum_IoU = 0
        sum_loss = 0

        with torch.no_grad():
            for images, masks, all_masks, style in test_loader:
                # Send tensors to cude
                images = images.to(device)
                masks = masks.to(device)
                all_masks = [x.to(device) for x in all_masks]

                # IoU/Loss on Image Level
                # outputs logits (the mean of the distribution)
                logits, output_dict, _ = model(images)
                pred_mask = (torch.sigmoid(logits)).ge(meta.masking_threshold)
                # Calculate the average ratio between foreground and background pixels
                loss_function = nn.BCEWithLogitsLoss(pos_weight=torch.tensor(2.0))
                sum_IoU += IoU(masks, pred_mask)
                sum_loss += loss_function(masks, logits)

                logit_distribution = output_dict["distribution"]
                n_pred_sigmoids = []
                for i in range(forward_passes):
                    logits = logit_distribution.sample()
                    n_pred_sigmoids.append(torch.sigmoid(logits))
                n_pred_sigmoids = torch.mean(
                    torch.stack(n_pred_sigmoids).float(), dim=0
                )

                # Log images, targets and predictions to wandb
                all_masks_heatmap = torch.mean(torch.stack(all_masks), dim=0)
                grid = make_image_grid_with_heatmaps(
                    images,
                    all_masks_heatmap,
                    n_pred_sigmoids,
                    required_padding=(0, 0, 0, 0),
                )
                if W:
                    wandb.log(
                        {
                            "Images during Validation": [
                                wandb.Image(
                                    grid, caption="Images, Targets, Predictions"
                                )
                            ]
                        }
                    )

            if W:
                wandb.run.summary[
                    "Loss on Validation Set (post Epoch)"
                ] = sum_loss / len(test_loader)
            if W:
                wandb.run.summary["IoU on Validation Set (post Epoch)"] = sum_IoU / len(
                    test_loader
                )

    else:
        print(
            f"\nCould not find trained model with the name {model.name}. Please check/train first and try again!\n"
        )


def evaluate_one_image(
    i, model, test_set, label_style, treshold_for_testing, forward_passes, meta, W=True
):
    """
    Evaluate the i-th image from the test set forward_passes times

    Args

    Returns
        Numpy array (HxW) with image
        List of Numpy array (HxW) with sigmoid predictions for the given number of forward passes
        List of Numpy array (HxW) with all annotations available for the image

    """
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    # if os.path.isfile(f"saved_models/{meta.directory_name}/{model.name}.pt"):
    #    model = torch.load(
    #        f"saved_models/{meta.directory_name}/{model.name}.pt")
    model.to(device)
    ssn_predictions = []
    with torch.no_grad():
        model.eval()
        image, mask, all_masks, _ = test_set[i]

        # Add a batch dimension
        image = torch.unsqueeze(image, 0)
        mask = torch.unsqueeze(mask, 0)
        # all_masks = [torch.unsqueeze(x, 0) for x in all_masks]

        # Send image to Cuda
        image = image.to(device)
        mask = mask.to(device)

        # outputs logits (the mean of the distribution)
        mean, output_dict, _ = model(image, label_style)
        logit_distribution = output_dict["distribution"]
        for i in range(forward_passes):
            logits = logit_distribution.sample()
            sigmoids = torch.sigmoid(logits)
            # pred_mask = (sigmoids).ge(treshold_for_testing) #Default is Treshold at 0.5 to get a binary prediction
            ssn_predictions.append(torch.squeeze(sigmoids).detach().cpu().numpy())

    image = torch.squeeze(image)
    if torch.Tensor.dim(image) > 2:
        image = image.permute((1, 2, 0))
    return (
        image.detach().cpu().numpy(),
        torch.squeeze(torch.sigmoid(mean)).detach().cpu().numpy(),
        ssn_predictions,
        [torch.squeeze(x).numpy() for x in all_masks],
        mask.detach().cpu().numpy(),
    )
    # else:
    #    print(
    #        f"\nCould not find trained model with the name {model.name}. Please check/train first and try again!\n")


def evaluate_one_image_for_ged(
    i,
    model,
    test_set,
    label_style_distribution,
    treshold_for_testing,
    forward_passes,
    meta,
    W=True,
):
    """
    Evaluate the i-th image from the test set forward_passes times

    Args

    Returns
        Numpy array (HxW) with image
        List of Numpy array (HxW) with sigmoid predictions for the given number of forward passes
        List of Numpy array (HxW) with all annotations available for the image

    """
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    # if os.path.isfile(f"saved_models/{meta.directory_name}/{model.name}.pt"):
    #    model = torch.load(
    #        f"saved_models/{meta.directory_name}/{model.name}.pt")
    model.to(device)
    ssn_predictions = []
    with torch.no_grad():
        model.eval()
        image, mask, all_masks, _ = test_set[i]

        # Add a batch dimension
        image = torch.unsqueeze(image, 0)
        mask = torch.unsqueeze(mask, 0)
        # all_masks = [torch.unsqueeze(x, 0) for x in all_masks]

        # Send image to Cuda
        image = image.to(device)
        mask = mask.to(device)

        for i in range(forward_passes):
            label_style = label_style_distribution.sample()
            label_style = label_style.unsqueeze(0).to(device)

            # outputs logits (the mean of the distribution)
            mean, output_dict, _ = model(image, label_style)
            logit_distribution = output_dict["distribution"]
            logits = logit_distribution.sample()
            sigmoids = torch.sigmoid(logits)
            # pred_mask = (sigmoids).ge(treshold_for_testing) #Default is Treshold at 0.5 to get a binary prediction
            ssn_predictions.append(torch.squeeze(sigmoids).detach().cpu().numpy())

    image = torch.squeeze(image)
    if torch.Tensor.dim(image) > 2:
        image = image.permute((1, 2, 0))
    return (
        image.detach().cpu().numpy(),
        torch.squeeze(torch.sigmoid(mean)).detach().cpu().numpy(),
        ssn_predictions,
        [torch.squeeze(x).numpy() for x in all_masks],
        mask.detach().cpu().numpy(),
    )`

The `evaluate_one_image_for_ged` allows you to hand over a label style distribution which should be a 'torch.distribution' object. You can then assume a label style distribution for the entire test set or fix the label style with a Dirac distribution and test on a subset of one label style. You can use the subset dataset class! 

Best, 
Kilian

**Comment by Michael-Li-Bryant** [comment]
Thanks a lot! Best wishes!
